### ISSUE TITLE
Need to send unencoded characters - Error: Request path contains unescaped characters

### ISSUE THREAD

**Issue by unreleased** [a85/POSTMan-Chrome-Extension#11206]
### Is there an existing issue for this?

- [X] I have searched the tracker for existing similar issues and I know that duplicates will be closed

### Describe the Issue

Hello,

Please allow me to give some background before I jump into this, I understand this issue is commonly reported and answered the same on all the Stackoverflow results.

For years I have been limited to using Insomnia's REST API because it's built using node-libcurl for requests rather than Node's HTTP module (Which I believe Postman is using based on the response I'm getting). The problem is, that Postman is much better built for handling my other use cases and is less buggy in general.

The error is "Error: Request path contains unescaped characters"

If you Google this, you'll see everyone is typically making beginner's mistakes not encoding something correctly. I am performing security research and I *want* to send unencoded space characters ` `, I absolutely and totally do not want it to be encoded and sent as `%20` or any other type of encoded method.

**Why do I want to do this?**

Certain servers behave unexpectedly when using unencoded space characters and it opens up the potential for a lot of abuse and accessing restricted routes which is why I need to be able to do this.

I believe the issue with sending unencoded spaces throwing an error is due to Node's core infrastructure but I'm not 100% certain. I really came here to ask: Does anyone have any sort of fix whatsoever for this? It can be as hacky, weird, strange or whatever verb you want to use but I really want to stop using other clients and stick to Postman. :(

### Steps To Reproduce

1. Disable automatic encoding
2. Put a space between any URL path and send the request

### Operating System

N/A

### Postman Version

N/A

### Postman Platform

Postman App

**Comment by postman-staff** [comment]
Hi @unreleased

Thank you for reporting this to us.
Unfortunately, there's no way to send a special unencoded character or space in the query parameters or URL. We also checked, and see that this feature is unavailable in most Rest clients, including CURL.

I believe this one is similar if not the same: https://github.com/postmanlabs/postman-app-support/issues/8912

We will make sure to have this tracked internally and you may subscribe to this thread to receive notifications once the feature is released. Thank you!

via @Raj-postman | Postman Support

**Comment by unreleased** [comment]
Hi @postman-support @Raj-postman 

This feature is available in cURL and libcurl for Node. That's how it's possible via Insomnia and the NPM package request-curl which I created specifically for this purpose.

**Comment by postman-staff** [comment]
Thank you for the update @unreleased. True, we were able to replicate this case and will have this reviewed.

via @Raj-postman | Postman Support

**Comment by codenirvana** [comment]
@unreleased I understand your use case and that's why we have the "Encode URL automatically" setting which you can turn off to disable automatic-encoding.

![Screenshot 2022-08-24 at 5 36 30 PM](https://user-images.githubusercontent.com/4865836/186414918-c0db5187-7a7a-4e72-bc70-2f20bdb9d172.png)

This works for characters in `[33, 255]` range but anything outside it must be encoded due to a [hard check](https://github.com/nodejs/node/blob/v18.7.0/lib/_http_client.js#L170..L174) in Node.js.

**Comment by unreleased** [comment]
Hi @codenirvana

I am attempting to send a request using an empty space character. Decoded is `%20`, but I don't want to send a decoded space, I want to send a space. When you turn off "Encode URL automatically" and attempt to send a space it will not work due to Node's hard check as you've mentioned.

Node's hard check follows RFC specifications but does not allow what's theoretically possible on the HTTP level. For my security research, I (and others) often need to step outside these guidelines to break and discover the inner workings of a system.

**Comment by codenirvana** [comment]
@unreleased I agree entirely with that, and Postman being a debugging client, must enable testing all these workflows.

We are currently using the `node:http` module, which has this limitation.
We've hit multiple such limitations in the past and had unblocked by either overriding the methods at the runtime or directly patching the Node.js source.

This might take a while, but we'll try to unblock this as well. Let me know if you find any workaround for this.

**Comment by unreleased** [comment]
I will have a look and get back if I discovered anything.

In the meantime, I have opened a feature request for Node itself to add an option to ignore the strict checking they're performing.

https://github.com/nodejs/node/issues/44381

**Comment by unreleased** [comment]
So, after further investigation, it is possible to send raw characters directly using the internal net module by writing the request on the socket directly. However, It would be an incredible amount of work for you guys to move all the HTTP(s) functionality away from the HTTP module as a solution to this unpopular issue.

I think a more appropriate solution would be patching the node source directly, however, that's outside of my expertise.
